### ISSUE TITLE
Instructions for enabling item categories

### ISSUE THREAD

**Issue by bf4** [minimul/qbo_api#41]
Sharing here since I'm not sure how to post an answer to an unasked question on the intuit site

## tl;dr 

to enable item categories in dev, when logged in to a sandbox company, go to https://sandbox.qbo.intuit.com/app/categorymigration and click 'Switch to Categories'

## Details

This is a new dev account using the OAuth 2.0 API 

I was unable to turn on categories in the sandbox by updating companyinfo or any obvious page in the web ui. 

In prod, I noticed that the URL for the categories page was `/app/category`, so I went to `https://sandbox.qbo.intuit.com/app/categories` and it redirected to `https://sandbox.qbo.intuit.com/app/categorymigration` and I clicked the link to 'Switch to Categories'. I was then able to go to `/app/categories` in the web ui to create categories, as well as create them via the API.

It appears the migration posted a CORS request to `https://sandbox.qbo.intuit.com/qbo50/neo/v1/company/193514655268479/companysetup/enableCategories` 
e.g.

 ```
 POST /item, payload: { Name: 'Transportation', Type: 'Category' }, params: { minorversion: 4} 
``` 

Before the enablecategories migration, the API returned I got a 400

 ```
[{:fault_type=>"ValidationFault", :error_code=>"500", :error_message=>"Unsupported Operation", :error_detail=>"Operation CREATE CATEGORY is not supported."}] 
``` 

afterwards 

``` 
{"Name"=>"Flowers", "Active"=>true, "FullyQualifiedName"=>"Transportation", "Type"=>"Category", "domain"=>"QBO", "sparse"=>false, "Id"=>"36", "SyncToken"=>"0", "MetaData"=>{"CreateTime"=>"2017-11-08T08:58:38-08:00", "LastUpdatedTime"=>"2017-11-08T08:58:38-08:00"}}
 ```

### Not a bug: Category of Type: Service

Weird note: a subsequent query returned with Type: Service instead of Type: Category. (Quickbooks support suggested I use a minor version 12 or greater, but I didn't go through the trouble to create a new company to check that)

```
{"Name"=>"Transportation", "Active"=>true, "FullyQualifiedName"=>"Transportation", "Type"=>"Service", "domain"=>"QBO", "sparse"=>false, "Id"=>"36", "SyncToken"=>"0", "MetaData"=>{"CreateTime"=>"2017-11-08T13:28:16-08:00", "LastUpdatedTime"=>"2017-11-08T13:31:39-08:00"} 
``` 

but I was still able to use it as a super category, 

e.g. 

```
 { "Name"=>"Hour", "Active"=>true, "SubItem"=>true, "ParentRef"=>{"value"=>"36", "name"=>"Transportation"}, "Level"=>1, "FullyQualifiedName"=>"Transportation:Hour", "Type"=>"Service", "domain"=>"QBO", "sparse"=>false, "Id"=>"37", "SyncToken"=>"0", "MetaData"=>{"CreateTime"=>"2017-11-08T13:31:18-08:00", "LastUpdatedTime"=>"2017-11-08T13:31:18-08:00"}} 
```

**Comment by minimul** [comment]
Awesome. Thanks for sharing.

**Comment by JohnSmall** [comment]
@bf4  "as well as create them via the API." can you expand on that please?

As far as I can see the only way to add categories via the api is to add them as items, which is crazy. 

Also I found that when I check the categories I've added via the web interface using the Intuit api explorer the Type='Category', but when I look at what comes through qbo_api they show up as 'Service'. Which you've already seen. This is very strange.

**Comment by JohnSmall** [comment]
@bf4 If you set the minor version by QboApi.minor_version = x, I found x=4 to work. Then you get the Category displaying as 'Category' instead of 'Service'

**Comment by JohnSmall** [comment]
@bf4  Ok, I found the docs on minor versions [Qbo docs on minor versions](https://developer.intuit.com/docs/00_quickbooks_online/2_build/20_explore_the_quickbooks_online_api/80_minor_versions) any minor version > 3 will bring back categories with type = 'Category' and items with SKU's

**Comment by bf4** [comment]
@JohnSmall I'm guess from your collective comments you've answered your own questions?

- that in the api a category is an item type
- that categories need to be enabled
- that the api only 'knows' categories in minor version 4+
- that it's ok if the api returns type: service

(The answers were all in the original post, right?)

**Comment by bf4** [comment]
I also wrote it up at http://www.benjaminfleischer.com/fragments/enabling_quickbooks_item_categories/

**Comment by minimul** [comment]
Thanks @bf4
